highfidelity/hifi
# Prevent thread issues with PickResults modified by Pointers

https://highfidelity.manuscript.com/f/cases/18506/RayPickResult-RayPickResult-bb1a07c73108db9911271cd4551f56eaddaef1a21fff1b8a7f0ecd0ccddb3469

This PR addresses a common crash involving PickResults. It makes the PickResult used by Pointers a separate copy so it can properly use it for its own calculations. I could not reproduce the crash, but was able to produce a similar crash by making the script engine process a lot of LaserPointers and PickResults, which will be used during this test.

## TEST PLAN
**Prerequisites**
- Have access to Backtrace, and have [this Backtrace page](https://highfidelity.sp.backtrace.io/dashboard/highfidelity/project/Interface/group/ddd602e3c1a920c656b6835eaecef986a54e0d0b9820db0af2089d483710f9bb) open to refresh/reference later.
- Have a way to know if your interface client was responsible for a crash on Backtrace, such as being logged in to a specific account.
- Have a program open to monitor the amount of memory Interface is using such as Task Manager. The test script will cause Interface to use a lot of memory, which is unrelated to the crash but may impede testing.
- Be in desktop mode, on a domain, with at least one small (<2 meter dimensions) entity near (within 50 meters) your avatar. You can verify it's an entity by going into the Create menu, selecting the object, and going into the properties menu and seeing a bunch of properties such as the type, model name, position, etc. It is not required for the object to be editable in any way, just that it exists.

**Test Steps**

Run these steps 30 times in master and this PR, or however many times it takes to get 3 repros on master if that is smaller. For both builds, record the number of crashes that match [this Backtrace page](https://highfidelity.sp.backtrace.io/dashboard/highfidelity/project/Interface/group/ddd602e3c1a920c656b6835eaecef986a54e0d0b9820db0af2089d483710f9bb), which shows a list of occurences of the crash the test script can reproduce.
1. In an appropriate domain and location, copy, paste, and enter the following script into the console: [laser-pointer-lock_stress-test.js](https://gist.github.com/sabrina-shanman/cd537e43939c40c63e987bff0d0dfc76). Do not run this script from "Running Scripts" because it will not unload unless interface settings are reset. The console can be accessed by enabling Settings>Developer Menu and clicking on Developer>Console.
1. Wait until Interface crashes. Monitor the memory usage by Interface on occasion. If Interface takes up too much memory (OS memory usage above 80%), just close interface and count that as a run.
1. Refresh the backtrace page and see if there is a new crash which matches your interface client. If so, record it in the repro count.

This test should pass if the crash repros on master three times but does not repro in this PR after the same number of runs.
